routing/per_route_options.go

```diff
@@ -0,0 +1,125 @@
+package routing
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"sync"
+
+	"github.com/cloudfoundry/cf-test-helpers/v2/cf"
+	"github.com/cloudfoundry/cf-test-helpers/v2/helpers"
+
+	. "github.com/cloudfoundry/cf-acceptance-tests/cats_suite_helpers"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/app_helpers"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/assets"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/random_name"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gexec"
+)
+
+var _ = RoutingDescribe("Per-Route Options", func() {
+	var (
+		appName        string
+		appId          string
+		instanceIds    [2]string
+		leastConnHost  string
+		roundRobinHost string
+	)
+
+	Context("when an app sets the loadbalancing algorithm", func() {
+		BeforeEach(func() {
+			appName = random_name.CATSRandomName("APP")
+			asset := assets.NewAssets()
+			leastConnHost = random_name.CATSRandomName("dora-lc")
+			roundRobinHost = random_name.CATSRandomName("dora-rr")
+			Expect(cf.Cf("push",
+				appName,
+				"-b", Config.GetRubyBuildpackName(),
+				"-m", DEFAULT_MEMORY_LIMIT,
+				"-p", asset.Dora,
+				"--var", fmt.Sprintf("domain=%s", Config.GetAppsDomain()),
+				"--var", fmt.Sprintf("name=%s", appName),
+				"--var", fmt.Sprintf("leastconnhost=%s", leastConnHost),
+				"--var", fmt.Sprintf("roundrobinhost=%s", roundRobinHost),
+				"-f", filepath.Join(asset.Dora, "man_route_opt_lb.yml"),
+			).Wait(Config.CfPushTimeoutDuration())).To(Exit(0))
+			appId = app_helpers.GetAppGuid(appName)
+			appInstanceRegex, _ := regexp.Compile("[[:alnum:]]{8}(-[[:alnum:]]{4}){4}")
+			for i := range 2 {
+				Eventually(func() bool {
+					fmt.Fprintf(GinkgoWriter, "Waiting for app instance %d to start...\n", i)
+					curl := helpers.Curl(Config, Config.Protocol()+leastConnHost+"."+Config.GetAppsDomain()+"/id", "-H", fmt.Sprintf("X-Cf-App-Instance: %s:%d", appId, i)).Wait()
+					id := string(curl.Out.Contents())
+					if appInstanceRegex.MatchString(id) {
+						instanceIds[i] = id
+						fmt.Fprintf(GinkgoWriter, "App instance %d has started. Instance ID: %s.\n", i, id)
+						return true
+					} else {
+						fmt.Fprintf(GinkgoWriter, "App instance %d is not ready yet. Response: %s, curl error: %s.\n", i, id, string(curl.Err.Contents()))
+						return false
+					}
+				}).Should(BeTrue())
+			}
+		})
+
+		AfterEach(func() {
+			app_helpers.AppReport(appName)
+			Expect(cf.Cf("delete", appName, "-f", "-r").Wait()).To(Exit(0))
+		})
+
+		Context("when it's set to round-robin", func() {
+			It("distributes requests evenly", func() {
+				doraUrl := fmt.Sprintf("%s%s.%s", Config.Protocol(), roundRobinHost, Config.GetAppsDomain())
+				var wg sync.WaitGroup
+				for i := 0; i < 10; i++ {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						defer GinkgoRecover()
+						helpers.Curl(Config, fmt.Sprintf("%s/delay/20", doraUrl), "-H", fmt.Sprintf("X-Cf-App-Instance: %s:0", appId))
```

Comment:
Wouldn't using the `X-Cf-App-Instance` header here necessitate that every request goes to that specific instance. Why not remove it?

Reply:
Also, for round-robin you don't need to use the `delay/20` endpoint right? Removing that will speed up this test run.

Reply:
In fact, why do you need this request spamming at all in this case. In order to test that round-robin is occurring you just need the following for loop to result in roughly each requests to each process, right?

Reply:
We are creating these requests because if there is no traffic to dora round-robin and least-connection will behave identical. So we enforce a situation where one instance has ten ongoing requests while the other has zero. That way we can observe the difference between round-robin and least-connection.

Re: wait group: If we did the wait group in line 85 we would only test the algorithm once those requests have finished and both instances are back to zero connections. I'll tighten the timings as the test should run fast but I don't really see a way to avoid it completely.